(PUP-9156) Add puppet ssl clean action

joshcooper opened the pull request:
Adds a `puppet ssl clean` action to delete a host's client cert, private key, public key, CSR and private key password file. Only the current host's files are deleted based on the `$certname` setting.

If the `--localca` option is specified, then the host's local copy of the CA certificate and CRL bundle are deleted.

joshcooper commented:
jenkins please test this on windows2012r2-64a with agent 6.0.x

joshcooper commented:
Reminder to externalize the strings for localization.

puppetcla commented:
CLA signed by all contributors.

Magisus commented on lib/puppet/application/ssl.rb:
```diff
@@ -18,7 +18,7 @@ def help
 USAGE
 -----
-puppet ssl <action> [--certname <NAME>]
+puppet ssl <action> [--certname <NAME>] [--local-ca]
```
I think this should be an underscore, since all the command names use underscores.

joshcooper replied:
yeah, we mostly cram multiple words into one (without dashes or underscores), eg `waitforcert`, `logdest`, `fromdate`, `sourceaddress`, `loadclasses`, `parseonly`, ...

We also use dashes:

```
$ git --no-pager grep ' option("--\w*-' lib/puppet/application/*.rb
lib/puppet/application/agent.rb:  option("--detailed-exitcodes") do |arg|
lib/puppet/application/agent.rb:  option("--job-id ID") do |arg|
lib/puppet/application/apply.rb:  option("--use-nodes")
lib/puppet/application/apply.rb:  option("--detailed-exitcodes")
lib/puppet/application/apply.rb:  option("--write-catalog-summary")
lib/puppet/application/device.rb:  option("--detailed-exitcodes") do |arg|
lib/puppet/application/face_base.rb:  option("--render-as FORMAT") do |format|
```

We do use underscores in two places:

```
$ git --no-pager grep ' option("--\w*_' lib/puppet/application/*.rb
lib/puppet/application/device.rb:  option("--to_yaml","-y")
lib/puppet/application/resource.rb:  option("--to_yaml","-y")
```

However all puppet settings can be specified on the command line e.g. `--dns_alt_names`, so we always use underscores there.

So I think using dashes makes sense for an action's option.... But I agree it's confusing having underscores and dashes... so what about shortening the action names, e.g. `puppet ssl download`, `puppet ssl submit`, `puppet ssl clean --local-ca`? The app would still accept `download_cert` and `submit_csr`, but the help would only mention the shorter version. Thoughts?

joshcooper replied:
Also how about `puppet ssl request` instead of `submit`?

Magisus commented on spec/lib/puppet/test_ca.rb:
```diff
@@ -49,6 +49,12 @@ def revoke(cert)
+      { private_key: host_key, csr: csr, cert: sign(csr, opts).content }
+    end
+
     def generate(name, opts)
```
Does it maybe make sense to just replace this method with your new one that returns more info?

joshcooper replied:
sure let me see what that entails.

joshcooper replied:
That was easier than expected. Updated.

justinstoller commented on lib/puppet/application/ssl.rb:
```diff
@@ -130,4 +138,16 @@ def verify(host)
+      path = Puppet[setting]
+      if Puppet::FileSystem.exist?(path)
+        Puppet::FileSystem.unlink(path)
+        puts "Deleted #{path}"
```
Why again can't we use Puppet.info here?

joshcooper replied:
The short answer is because `puppet ssl` uses `puts` elsewhere.

The long answer is... We could use `Puppet.{notice,info,etc}`. The good thing is colorization, and it would match `puppet cert clean` display:

```
$ bx puppet cert clean foobar
Warning: `puppet cert` is deprecated and will be removed in a future release.
   (location: /Users/josh/work/puppet/lib/puppet/application.rb:370:in `run')
Notice: Revoked certificate with serial 3
Notice: Removing file Puppet::SSL::Certificate foobar at '/Users/josh/.puppetlabs/etc/puppet/ssl/ca/signed/foobar.pem'
Notice: Removing file Puppet::SSL::Certificate foobar at '/Users/josh/.puppetlabs/etc/puppet/ssl/certs/foobar.pem'
Notice: Removing file Puppet::SSL::Key foobar at '/Users/josh/.puppetlabs/etc/puppet/ssl/private_keys/foobar.pem'
```
The bad thing IMO is it prepends the message type and the message may be excluded based on the current log level, though that could be a good thing for `Puppet.debug`...

I'm happy to switch things over to using puppet's logger, though I think we should file a separate ticket for that.

justinstoller replied:
Fair, I thought there was a point where we couldn't because the logger wasn't set up yet or something like that. I was just curious, I'm happy to defer to you regarding logging best practices (:

joshcooper replied:
ah, yeah you have to force the `:console` log destination, but that's easy for CLIs that always write to stdout/stderr

Magisus commented on lib/puppet/application/ssl.rb:
```diff
@@ -23,13 +23,13 @@ def help
   then the operation will fail.
 
-* download_cert:
+* download:
```
Shortening these is going to force us to make updates in the PE installer.

joshcooper replied:
It still accepts the old action names, the only difference is the help text...  but I can drop this commit?

joshcooper commented:
jenkins please test this on windows2012r2-64a with agent 6.0.x

joshcooper commented on lib/puppet/application/ssl.rb:
```diff
@@ -99,35 +113,51 @@ def verify(host)
+      path = Puppet[setting]
+      if Puppet::FileSystem.exist?(path)
+        Puppet::FileSystem.unlink(path)
+        puts _("Deleted %{path}") % { path: path }
```
Nit, I think elsewhere in puppet (and in `puppet cert clean`) we say "Removed ...". What does `puppetserver-ca` do?

Magisus replied:
We're saying "Cleaned files for..." in the CA CLI, as a collective, since the cleaning is done via an API call.

joshcooper commented:
I updated the following:

* Added `--localca` option for clean
* During a clean, error if we're on the CA, and we're about to clean the client cert used to perform CA commands via the REST API
* i18n strings
* Specify the type of file we're removing, eg certificate, private key, etc
* Use puppet's console logger, and add `--verbose`/`--debug` so it's possible to see other puppet debug messages, eg http pool
* Updated specs to use the `TestCa` helper

joshcooper commented:
jenkins please test this on windows2012r2-64a with agent 6.0.x

Magisus commented on lib/puppet/application/ssl.rb:
```diff
@@ -57,77 +87,111 @@ def main
     #   indent = "  " * (i+1)
-    #   puts "#{indent}#{issuer.subject.to_s}"
+    #   Puppet.notice "#{indent}#{issuer.subject.to_s}"
     # end
```
What is this commented block?

joshcooper replied:
It was some debugging to show the chain of certs on the agent (where the intermediate and root certs are what the agent downloads). Speaking of which does puppetserver require that the intermediate certs are in the CA bundle that the agent downloads, and are they included in the SSL handshake?

joshcooper replied:
Cool, makes sense. I think adding the intermediate certs and sending them in the SSL handshake is considered best practice. But it also adds overhead to SSL handshake: https://www.feistyduck.com/library/openssl-cookbook/online/apA-ssl-tls-deployment-best-practices.html

Magisus commented:
So doing some manual verification, I'm noticing that if you have a different certname configured in your `[master]` and your `[agent]` sections of `puppet.conf`, it doesn't seemt o be possible to clean up the master cert using this tool, now that we have removed the ability to specify a certname.

Magisus commented on lib/puppet/application/ssl.rb:
```diff
@@ -13,41 +16,68 @@ def help
   end
 
-  option('--certname NAME') do |arg|
-    options[:certname] = arg
```
Looks like we can get rid of this because the option exists by default, since it's a puppet setting?

joshcooper replied:
Yep exactly. I forgot it was a setting when I added it, then realized it wasn't needed

Magisus commented:
Tried this out manually, seems to work as expected. The workflow of using `puppetserver ca clean` followed by `puppet ssl clean` works well.

justinstoller commented on lib/puppet/application/ssl.rb:
```diff
@@ -57,77 +87,111 @@ def main
+  def clean(host)
+    if Puppet[:certname] == Puppet[:ca_server]
+      # make sure cert has been removed from the CA
+      cert = host.download_certificate_from_ca(Puppet[:certname])
```
This will either return the cert, nil (on 404) or raise if there was another issue. Do we need to handle any of those?
Most of them seem benign to let raise, though I don't know if we talked about this case (your CA's offline)?
```
puppet justin (mergejosh $) ‖ be puppet ssl clean
Error: Could not run: Failed to open TCP connection to localhost:8140 (Connection refused - connect(2) for "localhost" port 8140)
```

joshcooper replied:
yeah let me rescue that and print a better message. As it is, it's not clear why "ssl clean" would be making a network request.

joshcooper replied:
@justinstoller updated
